### ISSUE TITLE
[Question] Quality improvement using different generator parameters

### ISSUE THREAD

**Issue by george-roussos** [jik876/hifi-gan#11]
I was wondering if it is possible to further improve quality by adjusting the generators parameters in the MRF, like the hidden dimension or the kernel sizes. Ideally it would get rid of even more artifacts. Has anybody tried this?

**Comment by CookiePPP** [comment]
@george-roussos 

I've tested `upsample_initial_channel` of 768, [`discriminator_periods`](https://github.com/CookiePPP/VocoderComparisons/blob/57b49a930628c68adbcb49e93a831765a93c669e/repos/hifi-gan/config_v1_48Khz_multiGPU.json#L16) with lots of extra discriminators, using and not-using the 2 period with 44khz.
And I could only hear a consistent difference early into training.

---
What would you/we use/do to verify the results?
I find it hard to hear the artefacts in single speaker HiFi-GAN models already (thought this might just be my hearing going?), if you have something that consistently fails then that would make testing the params much easier.

**Comment by george-roussos** [comment]
It is that breathing problem I mentioned on the other thread, when constructing a TTS spectrogram. I just cannot seem to get rid of it. The voice itself sounds great and natural, especially after finetuning on ground truth mels from my Taco2 model. But whenever my speaker breathes, it sounds very metallic. I tried a run in which I used the ParallelWaveGAN generator and the HiFiGAN discriminators. The quality was much worse than HiFiGAN, however the breathing problem was completely eliminated. So I am kind of at a loss.

**Comment by CookiePPP** [comment]
@george-roussos 

~~https://96ad295d3358.ngrok.io/
Open this link and test the "Nancy" voice (at the top of the speaker list).
Does this speaker have the voice affect you're referring too?~~

---

Audio samples please?

**Comment by george-roussos** [comment]
Yes, sure. As I said my speaker has not given me consent so I cannot provide with actual voice, but for reference, here is the same breath mark from the same TTS spectrogram. The first breath mark is from HiFiGAN generator and the other one is from PWGAN generator.

https://drive.google.com/file/d/1jAaTtM6b2_QQ6EsEesYg4SVBwBglY7fE/view?usp=sharing

**Comment by CookiePPP** [comment]
Hmmm, I want to blame the tacotron2 model but I'm not sure. I don't use the same spectrogram config so I don't think I'll be doing any testing using my own models on your data. I guess good luck?

**Comment by george-roussos** [comment]
No worries mate, I appreciate you looking into it either way

**Comment by hdmjdp** [comment]
![图片](https://user-images.githubusercontent.com/26831425/99940974-301e9b80-2da8-11eb-81a0-d209b69b2b65.png)
have you encounter the high frequency artifact? @george-roussos @CookiePPP

**Comment by CookiePPP** [comment]
@hdmjdp 
I noticed it though I couldn't hear it clearly enough to worry me.
The reconstruction loss fmax wasn't configured so [`fmax_for_loss`](https://github.com/jik876/hifi-gan/blob/master/config_v1.json#L28) was still at 11025. I haven't tested with this value configured to 16000 or higher.

I did try the 2 period discriminator but that didn't improve things to my ears, either needed to wait longer for training or human ears can't tell periodicity above 11Khz with high accuracy.

**Comment by hdmjdp** [comment]
@thanks for repying. I trained 48k. So if fmax_for_loss=null, the fmax will be 24000.  I have trained 120k steps. How many steps have you trained to get a good wav?
below is librosa api:
"fmax      : float >= 0 [scalar]
        highest frequency (in Hz).
        If `None`, use `fmax = sr / 2.0`"

**Comment by george-roussos** [comment]
@hdmjdp I also did not encounter it enough, but I think it depends on the speaker in general. At around 800K steps for me the speech sounded great.

@CookiePPP could you please tell me how you implemented 2 period discriminator and more discriminators in general, as you mentioned?

**Comment by hdmjdp** [comment]
@george-roussos @CookiePPP  this is my config:
"upsample_rates": [10,5,3,2,2],
    "upsample_kernel_sizes": [20,11,7,4,4],
    "upsample_initial_channel": 384,
    "resblock_kernel_sizes": [3,7,11],
    "resblock_dilation_sizes": [[1,3,5], [1,3,5], [1,3,5]],

    "segment_size": 19200,
    "num_mels": 128,
    "num_freq": 1025,
    "n_fft": 4096,
    "hop_size": 600,
    "win_size": 2400,

    "sampling_rate": 48000,

    "fmin": 0,
    "fmax": 12000,
    "fmax_for_loss": null,

**Comment by CookiePPP** [comment]
@hdmjdp 
> I trained 48k. So if fmax_for_loss=null, the fmax will be 24000.

Hmmm, I'm not sure then.

---

To use custom periods. [line 164](https://github.com/jik876/hifi-gan/blob/master/models.py#L164-L173)
```python
class MultiPeriodDiscriminator(torch.nn.Module):
    def __init__(self):
        super(MultiPeriodDiscriminator, self).__init__()
        self.discriminators = nn.ModuleList([
            DiscriminatorP(2),
            DiscriminatorP(3),
            DiscriminatorP(5),
            DiscriminatorP(7),
            DiscriminatorP(11),
        ])
```
to
```python
class MultiPeriodDiscriminator(torch.nn.Module):
    def __init__(self, periods=None):
        super(MultiPeriodDiscriminator, self).__init__()
        self.periods = periods if periods is not None else [2, 3, 5, 7, 11]
        self.discriminators = nn.ModuleList()
        for period in self.periods:
            self.discriminators.append(DiscriminatorP(period))
```
and [line 33](https://github.com/jik876/hifi-gan/blob/master/train.py#L33)
```python
    mpd = MultiPeriodDiscriminator().to(device)
```
to
```python
    mpd = MultiPeriodDiscriminator(h["discriminator_periods"] if "discriminator_periods" in h.keys() else None).to(device)
```

**Comment by hdmjdp** [comment]
@CookiePPP 
Actually, I have use your period config, just write using hardcode in models.py. Such as below:
class MultiPeriodDiscriminator(torch.nn.Module):
    def __init__(self):
        super(MultiPeriodDiscriminator, self).__init__()
        self.discriminators = nn.ModuleList([
            #DiscriminatorP(2),
            DiscriminatorP(3),
            DiscriminatorP(5),
            DiscriminatorP(7),
            DiscriminatorP(11),
            DiscriminatorP(17),
            DiscriminatorP(23),
            #DiscriminatorP(37),
        ])

**Comment by CookiePPP** [comment]
@hdmjdp 
Then I don't know, dataset or training time I suppose.

**Comment by hdmjdp** [comment]
@CookiePPP 
I will train more steps. And another question,  whether more peroids can improve the quanlity of audio?

**Comment by LqNoob** [comment]
> ![图片](https://user-images.githubusercontent.com/26831425/99940974-301e9b80-2da8-11eb-81a0-d209b69b2b65.png) Have you encountered the high frequency artifact? @george-roussos @CookiePPP

Hi, @hdmjdp. 
I have the same problem, did you solve the high frequency artifact? 
Can increasing the number of training steps solve the problem?

**Comment by thecooltechguy** [comment]
Hi @LqNoob @hdmjdp , did you figure out how to solve the high frequency artifact? i'm facing it too, thanks!
